src/client/ws/tests.rs

```diff
@@ -1 +1,124 @@
+	let server = WebSocketTestServer::with_hardcoded_response("127.0.0.1:0".parse().unwrap(), String::new()).await;
+	let uri = to_ws_uri_string(server.local_addr());
+	let client = WsClient::new(&uri, WsConfig::default()).await.unwrap();
+	assert!(client.notification("notif", jsonrpc::Params::None).await.is_ok());
+}
+
+#[tokio::test]
+async fn method_not_found_works() {
+	let server =
+		WebSocketTestServer::with_hardcoded_response("127.0.0.1:0".parse().unwrap(), method_not_found(Id::Num(0_u64)))
+			.await;
+	let uri = to_ws_uri_string(server.local_addr());
+	let client = WsClient::new(&uri, WsConfig::default()).await.unwrap();
+	let response: Result<jsonrpc::JsonValue, Error> = client.request("say_hello", jsonrpc::Params::None).await;
+	assert_error_response(response, jsonrpc::ErrorCode::MethodNotFound, METHOD_NOT_FOUND.into());
+}
+
+#[tokio::test]
+async fn parse_error_works() {
+	let server =
+		WebSocketTestServer::with_hardcoded_response("127.0.0.1:0".parse().unwrap(), parse_error(Id::Num(0_u64))).await;
+	let uri = to_ws_uri_string(server.local_addr());
+	let client = WsClient::new(&uri, WsConfig::default()).await.unwrap();
+	let response: Result<jsonrpc::JsonValue, Error> = client.request("say_hello", jsonrpc::Params::None).await;
+	assert_error_response(response, jsonrpc::ErrorCode::ParseError, PARSE_ERROR.into());
+}
+
+#[tokio::test]
+async fn invalid_request_works() {
+	let server =
+		WebSocketTestServer::with_hardcoded_response("127.0.0.1:0".parse().unwrap(), invalid_request(Id::Num(0_u64)))
+			.await;
+	let uri = to_ws_uri_string(server.local_addr());
+	let client = WsClient::new(&uri, WsConfig::default()).await.unwrap();
+	let response: Result<jsonrpc::JsonValue, Error> = client.request("say_hello", jsonrpc::Params::None).await;
+	assert_error_response(response, jsonrpc::ErrorCode::InvalidRequest, INVALID_REQUEST.into());
+}
+
+#[tokio::test]
+async fn invalid_params_works() {
+	let server =
+		WebSocketTestServer::with_hardcoded_response("127.0.0.1:0".parse().unwrap(), invalid_params(Id::Num(0_u64)))
+			.await;
+	let uri = to_ws_uri_string(server.local_addr());
+	let client = WsClient::new(&uri, WsConfig::default()).await.unwrap();
+	let response: Result<jsonrpc::JsonValue, Error> = client.request("say_hello", jsonrpc::Params::None).await;
+	assert_error_response(response, jsonrpc::ErrorCode::InvalidParams, INVALID_PARAMS.into());
+}
+
+#[tokio::test]
+async fn internal_error_works() {
+	let server =
+		WebSocketTestServer::with_hardcoded_response("127.0.0.1:0".parse().unwrap(), internal_error(Id::Num(0_u64)))
+			.await;
+	let uri = to_ws_uri_string(server.local_addr());
+	let client = WsClient::new(&uri, WsConfig::default()).await.unwrap();
+	let response: Result<jsonrpc::JsonValue, Error> = client.request("say_hello", jsonrpc::Params::None).await;
+	assert_error_response(response, jsonrpc::ErrorCode::InternalError, INTERNAL_ERROR.into());
+}
+
+#[tokio::test]
+async fn subscription_works() {
```

Comment:
NOTE, this doesn't test that subscription is unsubscribed when the subscription is dropped.

The `MockedServer` needs then parse the responses and maintain the requests so I prefer to have the `MockedServer` as simple as possible for now.